nearcore/src/runtime/mod.rs

```diff
@@ -618,6 +618,9 @@ impl NightshadeRuntime {
             Some(Arc::new(StoreCompiledContractCache { store: self.store.clone() }));
         // Execute precompile_contract in parallel but prevent it from using more than half of all
         // threads so that node will still function normally.
+        //
+        // FIXME: a more appropriate way would be to limit the number of tasks executed by a global
+        // thread pool, than spawning new threads here.
```

Comment:
I am not sure we necessary want to limit the global pool. "foreground" work should be able to use all the cores, if required. It's rather that this one here is background work, and we don't want for it to occupy more than half of the cores. 

I don't think rayon has a direct API to just put a semaphore of some kind here. *but*, I think what we can do here is to use

https://docs.rs/rayon/latest/rayon/iter/plumbing/trait.Producer.html#method.min_len

to ask the rayon to divide the work at at most `available_parallelism()` / 2 chunks

Reply:
What the comment here suggests is exactly that – the work should still be submitted to the global pool, but should limit the number of jobs submitted (to a half of the available capacity), instead of setting up an entirely independent pool with its own threads and relying on the OS to do something sensible.

Reply:
I think this can largely be achieved with just a plain `windows` or a similar iterator transformer.

Reply:
Ah, I misunderstood that as "liminting *all* tasks", rather than "limiting these particular tasks"

FWIW, asked how one actually would do that here: https://gitter.im/rayon-rs/Lobby?at=62570d4799d94f5f0c9a07ad

Reply:
Yeah, good idea about windows (chunks perhaps? windows are overlapping). We might as well do this here, as, realisically, I don't expect anyone would come back to this code anytime soon

Reply:
Reading the chat logs, it seems like there's no other way to do this besides splitting the iterator into slices. One problem here is that this computation becomes much more susceptible to long tails and we lose the inherent balancing/stealing in the thread pool.

Reply:
I did something here. There's more code, but I think perf-wise and functional-requirements wise it is the ideal option.